articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,423 @@
+```
+
+*Note:* that the DDS specification has a known typo, where it says `-` are allowed, but the RTI documentation correctly lists `_` as allowed.
+
+Additionally, DDS has a hard limit on topic names of 256 characters, so an additional goal is to minimize the number of extra characters used when mapping from ROS to DDS names.
+
+## ROS 2 Topic and Service Name Constraints
+
+In this section an outline of the proposed constrains for ROS 2 topic and service names will be enumerated along with rationales where appropriate.
+
+For convenience here is a summary of all rules for topic and service names in ROS 2:
+
+- may contain alphanumeric characters (`[0-9|a-z|A-Z]`), underscores (`_`), or forward slashes (`/`)
+- may use balanced curly braces (`{}`) for substitutions
+- may start with a tilde (`~`), the private namespace substitution character
+- must not start with a numeric character (`[0-9]`)
+- must not contain any number of repeated underscores (`_`)
+- must not end with an underscore (`_`)
+- must not have an underscore (`_`) followed by a forward slash (`/`), i.e. `_/`
+- must not end with a forward slash (`/`)
+- must not contain any number of repeated forward slashes (`/`)
+- must separate a tilde (`~`) from the rest of the name with a forward slash (`/`), i.e. `~/foo` not `~foo`
+- must have balanced curly braces (`{}`) when used, i.e. `{sub}/foo` but not `{sub/foo` nor `/foo}`
+
+### Fully Qualified Names
+
+The topic and service name rules allow for some convenience syntax, which in some cases requires additional context to expand to the fully qualified name and then to the DDS equivalent name.
+For example, as outlined in further detail in the sections that follow, names may be relative (e.g. `foo` versus the absolute `/foo`), they may contain the private namespace substitution character (`~`), or arbitrary substitutions which are within the curly braces (`{}`) syntax.
+With context, each of these features can be expanded to some simple string to form the fully qualified name.
+Fully qualified names have these additional restrictions:
+
+- must start with a forward slash (`/`), i.e. they must be absolute
+- must not contain tilde (`~`) or curly braces (`{}`)
+
+Note that expanded substitutions must result in a valid name.
+For example, a name `foo_{bar}` is not valid if `bar` is expanded as `_baz`, because that would result in a fully qualified name of `foo__baz` which contains repeated underscores (`_`).
+
+### Uniform Resource Locators (URLs)
+
+Additionally, topic and service names can be represented in the [Uniform Resource Locator (URL)](https://en.wikipedia.org/wiki/Uniform_Resource_Locator) format to further disambiguate the resource name.
+A topic name may be preceded by a `rostopic://` scheme prefix, and a service name may be preceded by a `rosservice://` scheme prefix.
+For example, the absolute name `/foo` could also be represented as a topic with `rostopic:///foo` or as a service with `rosservice:///foo`.
```

Comment:
I believe they can be used at the same time, as the DDS topic names will differ. This is related to the on going discussion about "host-oriented" versus "content-oriented" addressing of services. What does "protocol" mean in this case? Which "namespace" are you referring to, the ROS namespace or the DDS topic namespaces (the ones separated by `__`) or maybe the prefix used in the DDS topic name?